### ISSUE TITLE
[Feature Request] add support for pushing to Gerrit (and/or support for custom actions)

### ISSUE THREAD

**Issue by rburgst** [ForkIssues/Tracker#123]
I would love be able to push to code review when using Gerrit. 

The way it works is that whenever you check out a branch then you push into a "review branch" which has the following reference

`refs/for/<branch-name>`

I am currently using source tree which doesn't have this feature either but at least it supports custom actions which basically allow me to do the same thing (and is probably way more flexible anyway).

![2017-10-11 at 09 00](https://user-images.githubusercontent.com/22971/31426229-bb9900ec-ae62-11e7-942b-58c34094dc83.png)

**Comment by DanPristupov** [comment]
Hi.

Is there any way how I recognise if repository uses Gerrit?

**Comment by rburgst** [comment]
Not really. Often gerrit uses a unique port but only if you are using ssh. So I wouldn't rely on that.

**Comment by clounie** [comment]
I wonder if this kind of custom action would be a nicer way to implement asks like #126?

**Comment by rburgst** [comment]
That would definitely solve that. I actually have a custom action to do a rebase including autostash. Super handy. It would be good to be able to configure whether the screen should refresh after execution of the action (I guess the default should be ‘true’

**Comment by DanPristupov** [comment]
I've never used custom actions. Can we discuss what you would like to see there?

*What things you would do with custom actions?*
I can't think of many right offhand:
* start build
* npm update (or any other package manager update)
What else?

*Should custom actions be saved globally or per repository?*

*We can introduce a scope parameter which will determine a target for which action will be available:*
* globally, in the application menu (i.e. for the repository)
* revision (in the revision context menu)
* tags (sidebar context menu)
* branches (sidebar context menu)

May be even patterns like `feature/*`, i.e. enable custom actions for items which match the pattern.

*We can use variables in the menu title like following `Push Gerrit '%branch-name%'`*
*What 'variables' are important?*
`%sha%`
`%commitMessage%`
`%refName%` - branch/remote branch/tag name
`%refLongName%` - same as previous, but will show 'origin/master' instead of 'master' for remote branches
`%refFullName%` - full reference name, can be useful for scripting

**Comment by rburgst** [comment]
> What things you would do with custom actions?

I push to specific review branches:

        git push orgin HEAD:refs/for/releases/release-2.0

I also often do a pull/rebase with autostash

        git fetch && git rebase --autostash

Doing things like building etc is certainly an option, but at least for myself I would probably do that in other tools. I focus on doing git related things in there but the possibilities are endless.

> Should custom actions be saved globally or per repository?

I only know it from source tree where its global, however, I occasionally struggle with different active projects requiring different release branches and its easy to get confused. 
So having it locally would be nice. 
On the other hand I use the pull/rebase with auto stash on all my repos.
So in short, the best would be both, I personally can get by with them being global (which makes it easier to manage probably as you dont have to hunt through all your repos and copy paste stuff).

> We can introduce a scope parameter which will determine a target for which action will be available:

that sounds nice, however, I guess you have to be careful to make it not overly complicated for the users.

> What 'variables' are important?

I guess having the details of the current selected branch, commit and possibly the currently selected file would be beneficial, see also the example screenshot from source tree.

**Comment by rburgst** [comment]
Any updates on this?

**Comment by rburgst** [comment]
Any update?

**Comment by DanPristupov** [comment]
@rburgst sorry, there are more prioritized things now, so no ETA yet. It's still in my todo though.

**Comment by DanPristupov** [comment]
Status update: I'm working on custom actions right now. Planning to introduce them in the next update.

**Comment by YesThatAllen** [comment]
I can see using this to rebase an employee's commits and push the result without claiming authorship. 

`git commit --author="First Last <initial@author.com>"`

**Comment by DanPristupov** [comment]
I released an early version of 1.0.67 which includes draft of the custom actions (CA). You can download it here: https://github.com/ForkIssues/Tracker/issues/287

Could you please try it give me some feedback?

A little bit about how to make it work:

Custom actions can be managed in Fork preferences:
![2018-05-21 at 19 53](https://user-images.githubusercontent.com/618115/40321845-aeca71c0-5d30-11e8-8715-02f1512b9072.png)

There are two targets for custom actions: repository and revision (do we also need branch and tag targets?).

Repository actions are listed in two places:
- Repository Menu
- 'Open In' dropdown on the toolbar
![2018-05-21 at 19 55](https://user-images.githubusercontent.com/618115/40321947-05c34a6a-5d31-11e8-9410-cfc4ac72d8a1.png)

Revision actions are available in commit context menu:
![2018-05-21 at 19 56](https://user-images.githubusercontent.com/618115/40321978-21683348-5d31-11e8-8268-87981f528245.png)

**Comment by rburgst** [comment]
Works pretty well. The only thing that I would need in addition is the ability to configure shortcut keys :)

**Comment by DanPristupov** [comment]
@rburgst great! Thank you for the feedback. I will create a separate issue for the shortcuts as I delay them, until I migrate Fork to multiple-windows system (otherwise I may implement key handling twice).

**Comment by clounie** [comment]
Will Fork still have tabs when you do that multi-window system?

**Comment by DanPristupov** [comment]
@clounie yes, sure. Similar to Finder/Safari.

**Comment by asos-rshev** [comment]
Hey, could you please share where custom actions are stored? Would like to sync them between my machines.

**Comment by asos-rshev** [comment]
@DanPristupov I like the separate json idea. Repository-wide actions would probably still need import/export, because usually you can make use of the same custom actions over multiple repos. Would you create a separate issue or you'd like me to do that?
